### PR TITLE
Added zstandard compression support with mutithreading support.

### PR DESCRIPTION
**Pull request by willyvmm**
Added support for Zstandard - Fast real-time compression algorithm http://www.zstd.net
https://github.com/facebook/zstd

prerequisites:
python-zstandard library https://github.com/indygreg/python-zstandard
pip3 install zstandard

There is also python-zstd implementation. It will not work, as this offer only basic support

    Features:
         - compression levels from 1 to 22 supported
            default value is 5

         - multithreaded compression support
            0: maximum cpu count
            -1..-99: maximum cpu count minus 1..99
            1..99: defined cpu usage
            default value is 1

     Usage: borg create -C zstd,level,threads ...

    Warning: This is NOT THREAD SAFE IMPLEMENTATION
      It allow only ONE python context to be created.
      It should work flawless as long as borg will call ONLY ONE compression job at time.

--
I'am not too familiar with Python/git, so please review my code. Thanks.

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
Thanks for the PR, but did you read our ticket / policy about adding compression algorithms?

**Comment by ThomasWaldmann** [comment]
Hmm, seems like this can work as a optional compression, so it is not required to have the additional python / binary dependencies on every platform / distribution, so it might be compatible with the policy.

File: src/borg/compress.pyx [hunk 1]
```diff
@@ -22,6 +22,21 @@ try:
 except ImportError:
     lzma = None
 
+import multiprocessing as mp
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: without "as mp" please. just search for "mp" to see why.

File: src/borg/compress.pyx [hunk 2]
```diff
@@ -22,6 +22,21 @@ try:
...
+
+"""
+python-zstandard
+https://github.com/indygreg/python-zstandard
+pip3 install zstandard
+There is also python-zstd implementation. It will not work as this offer only basic support
+"""
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: this needs to go into setup.py as a optional requirement.

File: src/borg/compress.pyx [hunk 3]
```diff
@@ -126,7 +141,8 @@ class LZ4(CompressorBase):
         osize = LZ4_compressBound(isize)
         buf = buffer.get(osize)
         dest = <char *> buf
-        osize = LZ4_compress_limitedOutput(source, dest, isize, osize)
+        with nogil:
+            osize = LZ4_compress_limitedOutput(source, dest, isize, osize)
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: why did you change this?

File: src/borg/compress.pyx [hunk 4]
```diff
@@ -149,7 +165,8 @@ class LZ4(CompressorBase):
             except MemoryError:
                 raise DecompressionError('MemoryError')
             dest = <char *> buf
-            rsize = LZ4_decompress_safe(source, dest, isize, osize)
+            with nogil:
+                rsize = LZ4_decompress_safe(source, dest, isize, osize)
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: why did you change this?

File: src/borg/compress.pyx [hunk 5]
```diff
@@ -217,6 +234,50 @@ class ZLIB(CompressorBase):
...
+    zstd compression / decompression binding
+
+    Features:
+         - compression levels from 1 to 22 supported
+            default value is 5
+         - multithreaded compression support
+            0: maximum cpu count
+            -1..-99: maximum cpu count minus 1..99
+            1..99: defined cpu usage
+            default value is 1
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: we'll have to discuss about multithreading after the basic issues are fixed.

File: src/borg/compress.pyx [hunk 6]
```diff
@@ -217,6 +234,50 @@ class ZLIB(CompressorBase):
...
+    def __init__(self, level=5, threads=1, **kwargs):
+        super().__init__(**kwargs)
+        self.level = level
+        self.threads = threads
+        if zstd is None:
+            raise ValueError('No zstd support found.')
+
+    def compress(self, data):
+        cctx = zstd.ZstdCompressor(level=self.level, threads=self.threads, write_content_size=True)
+        data = cctx.compress(bytes(data)) # not sure about that typecast but it works.
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: pep8: 2 blanks to the left of # and one to the right.

File: src/borg/compress.pyx [hunk 7]
```diff
@@ -344,6 +406,36 @@ class CompressionSpec:
...
+            elif count == 3:
+                maxcpu = mp.cpu_count() #get cpu count
+                level = int(values[1])
+                if not 0 <= level <= 22:
+                    raise ValueError
+
+                threads = int(values[2])
+
+                if threads == 0:
+                    threads=maxcpu # use maximum avaliable cpu's
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: pep8: blanks around operators, so it is " = ". also fix typos: "available" and "CPUs"

File: src/borg/compress.pyx [hunk 8]
```diff
@@ -344,6 +406,36 @@ class CompressionSpec:
...
+                elif threads < 0:
+                    threads = maxcpu + threads #threads is NEGATIVE ... remember !!
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: pep8

File: src/borg/compress.pyx [hunk 9]
```diff
@@ -344,6 +406,36 @@ class CompressionSpec:
...
+                    if threads < 1: # too less cpu's
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: not enough CPUs

File: src/borg/compress.pyx [hunk 10]
```diff
@@ -344,6 +406,36 @@ class CompressionSpec:
...
+                        raise ValueError
+                else:
+                    if threads > maxcpu: # too many cpu's
```
Comment (ThomasWaldmann) [inline src/borg/compress.pyx]: CPUs

**Comment by enkore** [comment]
Multithreaded compressors generally only make sense for larger chunks of data as the compression ratio tends to suffer and little speedup is attained for small chunks of data and multiple threads.

**Comment by willyvmm** [comment]
Multithreaded compression is related to this ticket: https://github.com/borgbackup/borg/issues/3115 but i can see it will not work as i imagined. 

I did PR to 1.1 fork because i need it to work :)

Thanks for all suggestions, i will rewrite the code and take all suggestions, and will  remove multithreading support.
I did some tests, and it seems that multiple threads was never spawned, zstd logic is using much bigger window and do not spawn more threads when it is working with standard size chunks. 
 
I think this ticket can be closed for now.

Best Regards 
Willy.

**Comment by ThomasWaldmann** [comment]
@willyvmm thanks!

Next PR against master, we'll discuss then whether we could adopt/backport that to 1.1.x.

**Comment by FabioPedretti** [comment]
@willyvmm any news? :)

**Comment by ThomasWaldmann** [comment]
@willyvmm did you do more work on this?

now as zstd made it even into the linux kernel, I guess we also could take it. :)

it would be perfect if it could be an optional dependency (so borg can still work, even if there is no libzstd [or however it is called] is available).

**Comment by ThomasWaldmann** [comment]
Hmm, I just started on this. Based on @willyvmm's stuff, addressing my own comments from above.

PR will come soon.
